### PR TITLE
Throttle replication traffic on moving partitions

### PR DESCRIPTION
**Pull request by aaronlevin-stripe**
Please see the design document on #150. Note there are a few departures
from that design doc worth pointing out:

1. `throttle` was renamed to `replication_throttle`
2. `replication_throttle` parameter was added to all endpoints that
result in partition movement.

Thank you to the primary author of this work, Josh Rosen.

This was tested on our QA cluster (running Kafka brokers on 2.2) by Nat Wilson.

Co-authored-by: Josh Rosen <joshrosen@stripe.com>
Co-authored-by: Nat Wilson <nat@stripe.com>

### CODE REVIEW

**Comment by aaronlevin-stripe** [comment]
@efeg two questions:

1. We've been working off the `migrate_to_kafka_2_0` branch. How do you want to handle the merge into `master`? I can do my best to cherry pick the changes and submit a second PR.
2. We're super happy to update documentation for this new feature, but didn't know how/where. Please advise!

Thanks! I look forward to your feedback.

**Comment by efeg** [comment]
@aaronlevin-stripe Thanks for the patch -- will start the review soon!

1. I suspect that the patch will not have much conflict with the `master`; hence, cherry-pick should not require much effort. Upon merging this patch, if you would submit a follow-up PR on `master`, that would be great!
2. [This PR](https://github.com/linkedin/cruise-control/pull/750) will enable contribution to the Wiki via a regular patch -- i.e. the patch will add the documentation under `cruise-control/docs/wiki`, and upon merging the patch, the changes will be reflected under Github Wiki tab.

**Comment by aaronlevin-stripe** [comment]
@efeg thanks for the feedback. I'll try and incorporate this right away. I may not be able to get started until friday, though.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/KafkaCruiseControl.java [hunk 1]
```diff
@@ -262,6 +266,8 @@ public void setUserTaskManagerInExecutor(UserTaskManager userTaskManager) {
    * @param excludedTopics Topics excluded from partition movement (if null, use topics.excluded.from.partition.movement)
    * @param replicaMovementStrategy The strategy used to determine the execution order of generated replica movement tasks
    *                                (if null, use default.replica.movement.strategies).
+   * @param replicationThrottle The replication throttle (bytes/second) to apply to both leaders and followers
+   *                            during the rebalance (if null, no throttling is applied).
```
[the remark this answers is not in the document]
Comment (aaronlevin-stripe) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/KafkaCruiseControl.java]: (done)

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java [hunk 2]
```diff
@@ -519,11 +531,15 @@ public boolean hasOngoingExecution() {
     private ExecutorState.State _state;
     private Set<Integer> _recentlyDemotedBrokers;
     private Set<Integer> _recentlyRemovedBrokers;
+    private Long _replicationThrottle;
```
[the remark this answers is not in the document]
Comment (aaronlevin-stripe) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java]: (done)

File: config/cruisecontrol.properties [hunk 3]
```diff
@@ -177,6 +177,9 @@ num.concurrent.intra.broker.partition.movements=2
 # The max number of leadership movement within the whole cluster at a given time.
 num.concurrent.leader.movements=1000
 
+# Default replica movement throttle. If not specified, movements unthrottled by default.
```
[the remark this answers is not in the document]
Comment (aaronlevin-stripe) [reply config/cruisecontrol.properties]: Having used this a few times in production, I agree :)

File: cruise-control/src/main/scala/com/linkedin/kafka/cruisecontrol/executor/ExecutorUtils.scala [hunk 4]
```diff
@@ -118,4 +119,20 @@ object ExecutorUtils {
...
+
+  def changeBrokerConfig(adminZkClient: AdminZkClient, brokerId: Int, config: Properties): Unit = {
+    adminZkClient.changeBrokerConfig(Some(brokerId), config)
+  }
+
+  def changeTopicConfig(adminZkClient: AdminZkClient, topic: String, config: Properties): Unit = {
+    adminZkClient.changeTopicConfig(topic, config)
+  }
+
+  def getAllBrokerIdsInCluster(kafkaZkClient: KafkaZkClient): java.util.List[java.lang.Integer] = {
```
[the remark this answers is not in the document]
Comment (aaronlevin-stripe) [reply cruise-control/src/main/scala/com/linkedin/kafka/cruisecontrol/executor/ExecutorUtils.scala]: (done)

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java [hunk 5]
```diff
@@ -744,6 +764,7 @@ private void interBrokerMoveReplicas() {
                  finishedDataMovementInMB, totalDataToMoveInMB,
                  totalDataToMoveInMB == 0 ? 100 : String.format(java.util.Locale.US, "%.2f",
                                                   (finishedDataMovementInMB * 100.0) / totalDataToMoveInMB));
+        throttleHelper.clearThrottles();
```
[the remark this answers is not in the document]
Comment (aaronlevin-stripe) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java]: Oof. Nice catch. This explains some behaviour we saw when running rebalances. OK, will look into fixing this.

[the remark this answers is not in the document]
Comment (aaronlevin-stripe) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java]: @efeg ok, I resolved this by having `waitForExecutionTaskToFinish()` return the tasks that have finished. Then I added a function to clear throttles for the tasks that have completed.

One part that's worth reviewing is that I had to figure out if a `ExecutionTask` was a valid "replication movement task", so there is a little function that calculates this. Mainly it checks that `task.brokerId() != -1` and that the state is not `IN_PROGRESS` or `STARTED`. I'm not sure how accurate this is or if there's a better way (ideally we'd have a little `TaskType` enum or class hierarchy or something to specify what kind of task each one is - if we did this then we could probably move the entire throttle code into the task itself, which is what I originally wanted to do - but this would only make sense for certain tasks).

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java [hunk 6]
```diff
@@ -744,6 +764,7 @@ private void interBrokerMoveReplicas() {
...
       }
       // After the partition movement finishes, wait for the controller to clean the reassignment zkPath. This also
       // ensures a clean stop when the execution is stopped in the middle.
```
[the remark this answers is not in the document]
Comment (aaronlevin-stripe) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java]: (done in most recent commit)

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/parameters/AddedOrRemovedBrokerParameters.java [hunk 7]
```diff
@@ -62,6 +64,10 @@ public boolean dryRun() {
     return _dryRun;
   }
 
+  public Long replicationThrottle() {
```
[the remark this answers is not in the document]
Comment (aaronlevin-stripe) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/parameters/AddedOrRemovedBrokerParameters.java]: (done)

**Comment by aaronlevin-stripe** [comment]
@efeg ok, I believe all PR feedback has been addressed. Please see my note (here: https://github.com/linkedin/cruise-control/pull/748#discussion_r299579766) about how I resolved the throttle removal.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (efeg) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/Executor.java]: Thanks for the update -- I think we may avoid using `task.brokerId() != -1`. Left a comment on this.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ReplicationThrottleHelper.java [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+              // the task should not be pending
+              task.state() != ExecutionTask.State.PENDING &&
+              // brokerId is -1 when moving replicas between brokers
+              task.brokerId() != -1;
+  }
+
+  // clear throttles for a specific list of execution tasks
+  void clearThrottles(List<ExecutionTask> completedTasks) {
+      if (throttlingEnabled()) {
+        LOG.info("Removing rebalance throttles from all brokers in the cluster");
```
[the remark this answers is not in the document]
Comment (aaronlevin-stripe) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ReplicationThrottleHelper.java]: (done)

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ReplicationThrottleHelper.java [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+        List<ExecutionProposal> completedProposals =
+                completedTasks
+                        .stream()
+                        // Filter for completed tasks related to replica movement
```
[the remark this answers is not in the document]
Comment (aaronlevin-stripe) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ReplicationThrottleHelper.java]: (done)

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ReplicationThrottleHelper.java [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+
+  // Determines if a candidate task is ready to have its throttles removed.
+  boolean shouldRemoveThrottleForTask(ExecutionTask task) {
+      return
+              // the task should not be in progress
+              task.state() != ExecutionTask.State.IN_PROGRESS &&
+              // the task should not be pending
+              task.state() != ExecutionTask.State.PENDING &&
+              // brokerId is -1 when moving replicas between brokers
+              task.brokerId() != -1;
```
[the remark this answers is not in the document]
Comment (aaronlevin-stripe) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ReplicationThrottleHelper.java]: (done)

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ReplicationThrottleHelper.java [hunk 11]
```diff
@@ -0,0 +1,231 @@
+/*
```
[the remark this answers is not in the document]
Comment (aaronlevin-stripe) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ReplicationThrottleHelper.java]: I fixed this file. Are there other files you think I should fix? I don't work in Java much so I think I had some default style stuff set in intellij. Wasn't sure what the project wanted.

**Comment by aaronlevin-stripe** [comment]
@efeg ok, all PR feedback addressed. Thanks for suggesting the `TaskType`! I must have forgot this existed from the last time I looked at this.

**Comment by aaronlevin-stripe** [comment]
@efeg before merging there is one edge case I want to think about quickly. Which is, if we complete a task and go to remove a throttle on a host, we should make sure there isn't another `IN_PROGRESS` task running for that host.

**Comment by aaronlevin-stripe** [comment]
(I'll see if I can update the tests for these edge cases too)

**Comment by aaronlevin-stripe** [comment]
@efeg ok, I've added a test and also cleaned up all the callsites and existing tests to use the incremental throttle removal.

there is a caveat, though. This does have a small change in behaviour. Previously we removed all throttles. Now we do leave some replica-related throttles around if they existed before and are not related to any of the replica movements. This is fine, but we also remove broker throttles. So there is a weird case where you might have a pre-existing throttle replica-throttle, but after performing a cruise-control-executed replica movement, the throttle on the broker that relates to the replica-throttle was removed.

I think this is _ok_ because we would have removed those throttles previously. I can think of a fix where we cache the existing "state of the throttles" before starting and then at the end "put the world back into an ok state" but I think we should leave that for a future PR if you're ok with that.

I'm about to go on vacation for a week but I'll try to check-in if you have any feedback.

**Comment by efeg** [comment]
@aaronlevin-stripe
Would you be interested in updating the documentation for this new feature?

We enabled contribution to the Wiki via a regular patch -- i.e. the patch will add the documentation under cruise-control/docs/wiki, and upon merging the patch, the changes will be reflected under Github Wiki tab.

**Comment by aaronlevin-stripe** [comment]
@efeg (back from vacation). yup, happy to write some documentation! Will try to get to this soon! Thanks for merging!! \o/!
